### PR TITLE
exported helper for determining if a linter is run at the expression/file level

### PR DESCRIPTION
**Pull request by MichaelChirico**
Towards #921. Pausing here for discussion before copying it over to more linters.

Still not very sure on the naming/signature here. There's a bit of a tension...

It's a bit weird to me to use `!is.null()` in the function and then `!is_tree_level()` -- the double negation feels odd.

But OTOH, using

```
if (is_tree_level(source_file, "complete", require_xml = TRUE))
```

instead feels worse -- (1) it feels better to declare "we want the expression-level tree; return if not" than to say "we don't want the complete-file tree (it's up to the reader's knowledge to remember the alternative is the expression-level tree)" (2) that approach would also require a tighter implementation, because `is.null(full_xml)` leaves three options for what's _non_-null. So we really want to filter based on what _is_ present, rather than what's not.

Another option would be to re-name the function, e.g. `not_tree_level()`, then we'd get

```
if (not_tree_level(source_file, "expression", require_xml = TRUE)) return(list())
```

but it doesn't stick with me right away.

### CODE REVIEW

**Comment by AshesITR** [comment]
tree doesn't seem to self-explanatory to me. What about `is_lint_level(level = c("expression", "file"), require_xml = FALSE)`?

**Comment by MichaelChirico** [comment]
> What about is_lint_level(level = c("expression", "file"), require_xml = FALSE)

I guess "lint level" is not immediately clear either (does it mean loose/strict linting maybe?), but is probably clear enough in actual usage.

Now I'm also questioning whether using `source_file` as the argument is also misleading -- `expression_metadata`?

**Comment by AshesITR** [comment]
The object isn't really expression metadata, is it?
I'm thinking of something with code or source in its name, but I agree `source_file` isn't ideal. Renaming is fine because the argument is positional anyway.

Maybe `source_code`? Reads fine to me, e.g.: `source_code$lines`, `source_code$file_lines`.

**Comment by MichaelChirico** [comment]
source_code sounds good to me! nice! source_expressions could also work because it ties nicely to the object's source (get_source_expressions)

**Comment by AshesITR** [comment]
Good catch, `source_expression` seems perfect.

**Comment by MichaelChirico** [comment]
@AshesITR ready for design review again. I think the main question is whether we want

```
if (!is_lint_level(source_expression, "expression"))
```

or

```
if (not_lint_level(source_expression, "expression"))
```

**Comment by AshesITR** [comment]
I prefer the `!` variant.
Would also rename `level = "complete"` to `level = "file"` in case we ever add static analysis capabilities across files (e.g. required for checking DESCRIPTION dependencies in packages).

Also, can we add the valid levels as a character to the default of level?
I'm fine with stopping if level isn't explicitly given, but I absolutely hate interfaces that don't tell me in the signature what valid string constants are (a recent example being rex' regex options).

Looks good with these changes I think.

**Comment by MichaelChirico** [comment]
> Also, can we add the valid levels as a character to the default of level?

The reason not to have it as a default is to force `!missing(level)`, and the valid values are in the first line of the function body...

OTOH I suppose we could use one of the levels as a default and then `stopifnot(!missing(level))`, WDYT?

**Comment by AshesITR** [comment]
Yeah that's what I meant.

**Comment by MichaelChirico** [comment]
OK will start applying this to the rest of the linters

**Comment by MichaelChirico** [comment]
This is ready for full review

**Comment by AshesITR** [comment]
```
Undocumented arguments in documentation object 'is_lint_level'
    'source_expression'
```

**Comment by MichaelChirico** [comment]
fixed the typo & a few other things, PTAL.

File: R/is_lint_level.R [hunk 1]
```diff
@@ -0,0 +1,21 @@
+#' Is this an expression- or a file-level source object?
+#'
+#' Helper for determining whether the current `source_expression` contains
+#'   the full file tree, or is just a single expression.
```
Comment (AshesITR) [inline R/is_lint_level.R:4]: Full file tree isn't really clear to me. Maybe write "contains all expressions in the current file, or just a single expression." instead?

File: R/is_lint_level.R [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+#'
+#' @param source_expression A parsed expression object, i.e., an element
+#'   of the object returned by [get_source_expressions()].
+#' @param level Which level of expression is being tested? `"expression"`
+#'   means an individual expression, while `"file"` means the full
+#'   source tree for this file.
+#' @param require_xml If `TRUE`, the XML equivalent of the parse tree
+#'   is required. Use this whenever writing a linter based on logic
+#'   expressed via Xpath.
```
Comment (AshesITR) [inline R/is_lint_level.R:13]: Typo: XPath has a capital P

File: R/is_lint_level.R [hunk 3]
```diff
@@ -0,0 +1,21 @@
+#' Is this an expression- or a file-level source object?
+#'
+#' Helper for determining whether the current `source_expression` contains
+#'   the full file tree, or is just a single expression.
+#'
+#' @param source_expression A parsed expression object, i.e., an element
+#'   of the object returned by [get_source_expressions()].
+#' @param level Which level of expression is being tested? `"expression"`
+#'   means an individual expression, while `"file"` means the full
+#'   source tree for this file.
```
Comment (AshesITR) [inline R/is_lint_level.R:10]: ditto.

**Comment by AshesITR** [comment]
The code changes look good to me. Just a few documentation suggestions.

File: NEWS.md [hunk 4]
```diff
@@ -174,6 +174,7 @@ function calls. (#850, #851, @renkun-ken)
 * `trailing_whitespace_linter()` ignores trailing whitespace in strings by default. 
   This can be disabled using `allow_in_strings = FALSE` (#1045, @AshesITR)
 * Moved the default lintr cache directory from `~/.R/lintr_cache` to `R_user_dir("lintr", "cache")`. Note that this major version update invalidated the old cache anyway, so it can be safely deleted. (#1062, @AshesITR)
+* New `is_lint_level()` helper for readably explaining which type of expression is required for a custom linter. Some linters are written to either require the full file's parse tree (for example, `single_quotes_linter()`. Others only need single expressions, which is more cache-friendly (most linters are written this way to leverage caching). (#921, @michaelchirico)
```
Comment (AshesITR) [inline NEWS.md:177]:
```suggestion
* New `is_lint_level()` helper for readably explaining which type of expression is required for a custom linter. Some linters are written to either require the full file's parse tree (for example, `single_quotes_linter()`). Others only need single expressions, which is more cache-friendly (most linters are written this way to leverage caching). (#921, @michaelchirico)
```

**Review by AshesITR — COMMENTED** [review]
mini typo. rest LGTM. Thanks!

**Review by AshesITR — APPROVED** [review]

**Pull request merged**